<a href="https://colab.research.google.com/github/Sana-0511/Document_Summarization-/blob/main/TF_IDF.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
# Import necessary libraries
import nltk  # Natural Language Toolkit for NLP tasks
from sklearn.feature_extraction.text import TfidfVectorizer  # TF-IDF vectorizer
from sklearn.metrics.pairwise import cosine_similarity  # Cosine similarity metric

# Download the punkt tokenizer for sentence tokenization
nltk.download('punkt')

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.


True

In [ ]:
# Install the required utility
!apt-get install wget

# Import the necessary libraries
import zipfile
import os

# Provide the Google Drive link
google_drive_link = 'https://drive.google.com/uc?id=1FV5aYhpvl6NFrH7aYzEVg8pmO_yokYoI'

# Specify the output file name
zip_file = 'example.zip'

# Download the zip file
!wget --no-check-certificate $google_drive_link -O $zip_file

# Create a directory to extract the contents
extracted_dir = 'extracted_content'
os.makedirs(extracted_dir, exist_ok=True)

# Extract the contents of the zip file
with zipfile.ZipFile(zip_file, 'r') as zip_ref:
    zip_ref.extractall(extracted_dir)

# List the files in the extracted directory
extracted_files = os.listdir(extracted_dir)
print("Files extracted:", extracted_files)

Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
wget is already the newest version (1.21.2-2ubuntu1).
0 upgraded, 0 newly installed, 0 to remove and 24 not upgraded.
--2024-01-13 11:25:02--  https://drive.google.com/uc?id=1FV5aYhpvl6NFrH7aYzEVg8pmO_yokYoI
Resolving drive.google.com (drive.google.com)... 74.125.126.102, 74.125.126.100, 74.125.126.138, ...
Connecting to drive.google.com (drive.google.com)|74.125.126.102|:443... connected.
HTTP request sent, awaiting response... 303 See Other
Location: https://drive.usercontent.google.com/download?id=1FV5aYhpvl6NFrH7aYzEVg8pmO_yokYoI [following]
--2024-01-13 11:25:02--  https://drive.usercontent.google.com/download?id=1FV5aYhpvl6NFrH7aYzEVg8pmO_yokYoI
Resolving drive.usercontent.google.com (drive.usercontent.google.com)... 142.250.136.132, 2607:f8b0:4001:c34::84
Connecting to drive.usercontent.google.com (drive.usercontent.google.com)|142.250.136.132|:443... connected.
HTTP request sent, aw

In [ ]:
def read_file(file_path):
    # Function to read the content of a file
    with open(file_path, 'r', encoding='utf-8') as file:
        text = file.read()
    return text

In [ ]:
def sentence_extraction(text, num_sentences=5):
    # Tokenize the text into sentences using the punkt tokenizer
    sentences = nltk.sent_tokenize(text)

    # Create TF-IDF vectorizer
    tfidf_vectorizer = TfidfVectorizer()

    # Fit and transform the sentences using TF-IDF to obtain a numerical representation
    tfidf_matrix = tfidf_vectorizer.fit_transform(sentences)

    # Calculate cosine similarity between sentences to measure their similarity
    similarity_matrix = cosine_similarity(tfidf_matrix, tfidf_matrix)

    # Identify the most important sentences based on their similarity scores
    sentence_scores = [(i, sum(similarity_matrix[i])) for i in range(len(sentences))]
    sentence_scores = sorted(sentence_scores, key=lambda x: x[1], reverse=True)

    # Select the top N sentences for the summary
    summary_sentences = [sentences[i] for i, _ in sentence_scores[:num_sentences]]

    return ' '.join(summary_sentences)

In [ ]:
if __name__ == "__main__":
    # Specify the path to your text file
    file_path = '/content/extracted_content/example/msft.txt'

    # Read the file content
    text_content = read_file(file_path)

    # Set the number of sentences you want in the summary
    num_summary_sentences = 3

    # Generate the summary using TF-IDF sentence extraction
    summary = sentence_extraction(text_content, num_summary_sentences)

    # Print the generated summary
    print("\nSummary:\n", summary)



Summary:
 As part of the program, the Redmond giant which wants to expand its reach and is planning to build a strong developer ecosystem in India with the program will set up the core AI infrastructure and IoT Hub for the selected campuses. The company will provide AI development tools and Azure AI services such as Microsoft Cognitive Services, Bot Services and Azure Machine Learning.According to Manish Prakash, Country General Manager-PS, Health and Education, Microsoft India, said, "With AI being the defining technology of our time, it is transforming lives and industry and the jobs of tomorrow will require a different skillset. In an attempt to build an AI-ready workforce, Microsoft announced Intelligent Cloud Hub which has been launched to empower the next generation of students with AI-ready skills.


In an attempt to build an AI-ready workforce, Microsoft announced Intelligent Cloud Hub which has been launched to empower the next generation of students with AI-ready skills
Envisioned as a three-year collaborative program, Intelligent Cloud Hub will support around 100 institutions with AI infrastructure, course content and curriculum, developer support, development tools and give students access to cloud and AI services
As part of the program, the Redmond giant which wants to expand its reach and is planning to build a strong developer ecosystem in India with the program will set up the core AI infrastructure and IoT Hub for the selected campuses
The company will provide AI development tools and Azure AI services such as Microsoft Cognitive Services, Bot Services and Azure Machine Learning.According to Manish Prakash, Country General Manager-PS, Health and Education, Microsoft India, said, "With AI being the defining technology of our time, it is transforming lives and industry and the jobs of tomorrow will require a different skillset
This will require more collaborations and training and working with AI
That’s why it has become more critical than ever for educational institutions to integrate new cloud and AI technologies
The program is an attempt to ramp up the institutional set-up and build capabilities among the educators to educate the workforce of tomorrow." The program aims to build up the cognitive skills and in-depth understanding of developing intelligent cloud connected solutions for applications across industry
Earlier in April this year, the company announced Microsoft Professional Program In AI as a learning track open to the public
The program was developed to provide job ready skills to programmers who wanted to hone their skills in AI and data science with a series of online courses which featured hands-on labs and expert instructors as well
This program also included developer-focused AI school that provided a bunch of assets to help build AI skills.

Summary: As part of the program, the Redmond giant which wants to expand its reach and is planning to build a strong developer ecosystem in India with the program will set up the core AI infrastructure and IoT Hub for the selected campuses. The company will provide AI development tools and Azure AI services such as Microsoft Cognitive Services, Bot Services and Azure Machine Learning.According to Manish Prakash, Country General Manager-PS, Health and Education, Microsoft India, said, "With AI being the defining technology of our time, it is transforming lives and industry and the jobs of tomorrow will require a different skillset. In an attempt to build an AI-ready workforce, Microsoft announced Intelligent Cloud Hub which has been launched to empower the next generation of students with AI-ready skills.

In [ ]:
!pip install rouge
from rouge import Rouge

# Your original text and generated summary
original_text = text_content
generated_summary = summary

# Initialize the Rouge object
rouge = Rouge()

# Calculate ROUGE scores
scores = rouge.get_scores(generated_summary, original_text)

# Print the scores
print("ROUGE Scores:")
print(scores)

ROUGE Scores:
[{'rouge-1': {'r': 0.5138121546961326, 'p': 1.0, 'f': 0.6788321123040654}, 'rouge-2': {'r': 0.42517006802721086, 'p': 0.9920634920634921, 'f': 0.5952380910380953}, 'rouge-l': {'r': 0.5138121546961326, 'p': 1.0, 'f': 0.6788321123040654}}]


ROUGE-1:

Precision (p): 1.000
Recall (r): 0.514
F1-score (f): 0.679

ROUGE-2:

Precision (p): 0.992
Recall (r): 0.425
F1-score (f): 0.595

ROUGE-L:

Precision (p): 1.000
Recall (r): 0.514
F1-score (f): 0.679

These scores suggest very high precision in terms of matching unigrams (ROUGE-1) and capturing the longest common subsequence (ROUGE-L). The recall values are also relatively high, indicating that the generated summary effectively covers a significant portion of the important content from the original text. The F1-score, balancing precision and recall, reflects a strong overall performance, suggesting that the generated summary is both accurate and comprehensive.

In [ ]:
if __name__ == "__main__":
    # Specify the path to your text file
    file_path = '/content/extracted_content/example/computer.txt'

    # Read the file content
    text_content = read_file(file_path)

    # Set the number of sentences you want in the summary
    num_summary_sentences = 3

    # Generate the summary using TF-IDF sentence extraction
    summary = sentence_extraction(text_content, num_summary_sentences)

    # Print the generated summary
    print("\nSummary:\n", summary)



Summary:
 This term may also refer to a group of computers that are linked and function together, such as a computer network or computer cluster. Modern digital electronic computers can perform generic sets of operations known as programs. A broad range of industrial and consumer products use computers as control systems.


A computer is a machine that can be programmed to carry out sequences of arithmetic or logical operations automatically. Modern digital electronic computers can perform generic sets of operations known as programs. These programs enable computers to perform a wide range of tasks. A computer system is a nominally complete computer that includes the hardware, operating system, and peripheral equipment needed and used for full operation. This term may also refer to a group of computers that are linked and function together, such as a computer network or computer cluster. A broad range of industrial and consumer products use computers as control systems. Simple special-purpose devices like microwave ovens and remote controls are included, as are factory devices like industrial robots and computer-aided design, as well as general-purpose devices like personal computers and mobile devices like smartphones. Computers power the Internet, which links billions of other computers and users.

Summary:
 This term may also refer to a group of computers that are linked and function together, such as a computer network or computer cluster. Modern digital electronic computers can perform generic sets of operations known as programs. A broad range of industrial and consumer products use computers as control systems.

In [ ]:
from rouge import Rouge

# Your original text and generated summary
original_text = text_content
generated_summary = summary

# Initialize the Rouge object
rouge = Rouge()

# Calculate ROUGE scores
scores = rouge.get_scores(generated_summary, original_text)

# Print the scores
print("ROUGE Scores:")
print(scores)

ROUGE Scores:
[{'rouge-1': {'r': 0.4270833333333333, 'p': 1.0, 'f': 0.5985401417912516}, 'rouge-2': {'r': 0.3333333333333333, 'p': 0.9591836734693877, 'f': 0.4947368382775624}, 'rouge-l': {'r': 0.4270833333333333, 'p': 1.0, 'f': 0.5985401417912516}}]


ROUGE-1:

Precision (p): 1.000
Recall (r): 0.427
F1-score (f): 0.599

ROUGE-2:

Precision (p): 0.959
Recall (r): 0.333
F1-score (f): 0.495

ROUGE-L:

Precision (p): 1.000
Recall (r): 0.427
F1-score (f): 0.599

These scores suggest high precision in terms of matching unigrams (ROUGE-1) and capturing the longest common subsequence (ROUGE-L). However, the recall values are moderate, indicating that while the generated summary contains correct unigrams and overlaps well with the reference in terms of the longest common subsequence, there may be some content from the original text that is not covered. The F1-score provides a balance between precision and recall, reflecting a good overall performance.

In [ ]:
if __name__ == "__main__":
    # Specify the path to your text file
    file_path = '/content/extracted_content/example/space.txt'

    # Read the file content
    text_content = read_file(file_path)

    # Set the number of sentences you want in the summary
    num_summary_sentences = 3

    # Generate the summary using TF-IDF sentence extraction
    summary = sentence_extraction(text_content, num_summary_sentences)

    # Print the generated summary
    print("\nSummary:\n", summary)



Summary:
 But this is deeply controversial, because Einstein’s theory predicts the existence of a so-called singularity at the centre of black holes – a state of infinite gravity which would prevent anything from passing through to the white hole on the other side. But unlike black holes, there’s no consensus about whether white holes exist, or how they’d be formed.They are predicted by Einstein’s theory of gravity, and are most often mentioned in the context of ‘wormholes’, in which a black hole acts as the entry point to a tunnel through space and time, ending in a white hole somewhere else in the Universe. But then, quantum effects occurring around the surface of the black hole halt further collapse to a singularity, and instead begin to gradually turn the black hole into a white hole that’s spewing out the original star matter again.


A white hole is a bizarre cosmic object which is intensely bright, and from which matter gushes rather than disappears. In other words, it’s the exact opposite of a black hole. But unlike black holes, there’s no consensus about whether white holes exist, or how they’d be formed.They are predicted by Einstein’s theory of gravity, and are most often mentioned in the context of ‘wormholes’, in which a black hole acts as the entry point to a tunnel through space and time, ending in a white hole somewhere else in the Universe. But this is deeply controversial, because Einstein’s theory predicts the existence of a so-called singularity at the centre of black holes – a state of infinite gravity which would prevent anything from passing through to the white hole on the other side. The process starts when an old massive star collapses under its own weight and forms a black hole. But then, quantum effects occurring around the surface of the black hole halt further collapse to a singularity, and instead begin to gradually turn the black hole into a white hole that’s spewing out the original star matter again. The process is mind-bendingly slow, though, so we may be in for a very long wait to find out if white holes really exist.

Summary:
 But this is deeply controversial, because Einstein’s theory predicts the existence of a so-called singularity at the centre of black holes – a state of infinite gravity which would prevent anything from passing through to the white hole on the other side. But unlike black holes, there’s no consensus about whether white holes exist, or how they’d be formed.They are predicted by Einstein’s theory of gravity, and are most often mentioned in the context of ‘wormholes’, in which a black hole acts as the entry point to a tunnel through space and time, ending in a white hole somewhere else in the Universe. But then, quantum effects occurring around the surface of the black hole halt further collapse to a singularity, and instead begin to gradually turn the black hole into a white hole that’s spewing out the original star matter again.

In [ ]:
!pip install rouge
from rouge import Rouge

# Your original text and generated summary
original_text = text_content
generated_summary = summary

# Initialize the Rouge object
rouge = Rouge()

# Calculate ROUGE scores
scores = rouge.get_scores(generated_summary, original_text)

# Print the scores
print("ROUGE Scores:")
print(scores)

ROUGE Scores:
[{'rouge-1': {'r': 0.6911764705882353, 'p': 1.0, 'f': 0.8173912995145558}, 'rouge-2': {'r': 0.6700507614213198, 'p': 0.9924812030075187, 'f': 0.7999999951880625}, 'rouge-l': {'r': 0.6911764705882353, 'p': 1.0, 'f': 0.8173912995145558}}]


ROUGE-1:

Precision (p): 1.000
Recall (r): 0.691
F1-score (f): 0.817

ROUGE-2:

Precision (p): 0.992
Recall (r): 0.670
F1-score (f): 0.800

ROUGE-L:

Precision (p): 1.000
Recall (r): 0.691
F1-score (f): 0.817

These scores indicate extremely high precision in terms of matching unigrams (ROUGE-1) and capturing the longest common subsequence (ROUGE-L). The recall values are also very high, suggesting that the generated summary effectively covers a substantial portion of the important content from the original text. The F1-score, balancing precision and recall, reflects an outstanding overall performance, indicating that the generated summary is both highly accurate and comprehensive.

## REFERENCES

https://www.geeksforgeeks.org/understanding-tf-idf-term-frequency-inverse-document-frequency/

https://towardsdatascience.com/tf-idf-for-document-ranking-from-scratch-in-python-on-real-world-dataset-796d339a4089

https://towardsdatascience.com/text-summarization-using-tf-idf-e64a0644ace3